src/lib/data/planEntities.ts

```diff
@@ -0,0 +1,236 @@
+ * validate the consistency of data, and return it in a useful format
+ */
+export const getAndValidateAllPlanEntities = async ({
+  database,
+  planId,
+  governingEntities,
+  prototypes,
+}: {
+  database: Database;
+  planId: PlanId;
+  /**
+   * A map from governing entities that include the version to be used when
+   * generating references
+   */
+  governingEntities: MapOfGoverningEntities;
+  version: 'latest';
+  /**
+   * A map of prototypes that **must** include the prototype for governing
+   * entities of this plan.
+   *
+   * Usually the full list of prototypes for a plan is passed in.
+   */
+  prototypes: Map<
+    EntityPrototypeId,
+    InstanceDataOfModel<Database['entityPrototype']>
+  >;
+}): Promise<ValidatedPlanEntities> => {
+  const planEntities = await database.planEntity.find({
+    where: {
+      planId,
+    },
+  });
+  const planEntityIDs = new Set(planEntities.map((pe) => pe.id));
+
+  const pevs = await database.planEntityVersion.find({
+    where: (builder) =>
+      builder
+        .whereIn('planEntityId', [...new Set(planEntities.map((pe) => pe.id))])
+        .andWhere('latestVersion', true),
+  });
+  const pevsByPlanEntityId = organizeObjectsByUniqueProperty(
+    pevs,
+    'planEntityId'
+  );
+
+  const entitiesAssociation = await database.entitiesAssociation.find({
+    where: (builder) =>
+      builder
+        .whereIn('childId', [...new Set(planEntities.map((pe) => pe.id))])
+        .andWhere('parentType', 'governingEntity')
+        .andWhere('childType', 'planEntity'),
+  });
+  const eaByChildId = organizeObjectsByUniqueProperty(
+    entitiesAssociation,
+    'childId'
+  );
+
+  const result: ValidatedPlanEntities = new Map();
+
+  for (const planEntity of planEntities) {
+    const planEntityVersion = pevsByPlanEntityId.get(planEntity.id);
+    if (!planEntityVersion) {
+      throw new Error(
+        `Missing plan entity version for planEntity ${planEntity.id}`
+      );
+    }
+
+    const prototype = prototypes.get(planEntity.entityPrototypeId);
+    if (!prototype) {
+      throw new Error(`Missing prototype for planEntity ${planEntity.id}`);
+    }
+
+    const refAndType = getCustomReferenceAndTypeForPlanEntity({
+      planEntity,
+      planEntityVersion,
+      prototypes,
+      entityAssociations: eaByChildId,
+      governingEntities,
+    });
+
+    const entityDetails: ValidatedPlanEntity = {
+      id: planEntity.id,
+      type: refAndType.type,
+      customRef: refAndType.customRef,
+      description: null,
+      supports: [],
+      governingEntity: eaByChildId.get(planEntity.id)?.parentId || null,
+    };
+
+    // Use entity details if possible
+    if (planEntityVersion.value) {
+      // Able to inspect entity details
+      entityDetails.description = planEntityVersion.value.description;
+      const supportingPlanEntityIds = planEntityVersion.value.support
+        .map((s) => s.planEntityIds)
+        .flat();
```

Comment:
`Array.prototype.flatMap()` "is identical to a `map()` followed by a `flat()` of depth 1, but slightly more efficient than calling those two methods separately".

And `Array.prototype.flat()` accepts `depth` parameter, which defaults to 1 when not supplied.
```suggestion
        .flatMap((s) => s.planEntityIds);
```

Reply:
Nice!